### ISSUE TITLE
The Session works on the Request and Response instances before the ComponentChain

### ISSUE THREAD

**Issue by albe** [neos/flow-development-collection#514]
The Session is implemented such that it keeps references to the `Request`/`Response` instances gotten from the active `RequestHandler` on initialization, **before** the ComponentChain has been invoked (bootstrap scripts).

Since the `ComponentChain` is supposed to possibly **replace** both instances, this can lead to the Session not working any longer.

The main reason this went by unnoticed is that the Session currently only uses the `Request` to get the session cookie and that no Http Component exists, that replaces the Response object.

**Comment by albe** [comment]
I see following possible solutions:

- make the Session do all work lazily (still not 100% safe)
- make Session a Component itself. That would somehow require the ComponentChain to work inward&outward though, like the PHP middleware PSR, i.e. be able to run *around* later components instead of just *before*, as it needs to *set* the session cookie on the `Response` as late as possible

**Comment by bwaidelich** [comment]
Can't we solve it similar to #499 so that the `Session` gets an instance of the `ComponentChain` rather than `Request`/`Response` directly?

**Comment by albe** [comment]
Session is initialized/started in the bootup scripts *before* the ComponentChain. That's actually the big issue, so ComponentChain/ComponentContext isn't even available at that point.

**Comment by bwaidelich** [comment]
> Session is initialized/started in the bootup scripts before the ComponentChain.

Right, but is there some intrinsic reason for this or can the order be changed?

**Comment by albe** [comment]
That's what I'd like to discuss/elaborate :)

**Comment by bwaidelich** [comment]
> That's what I'd like to discuss/elaborate :)
I'm fine with it ;)

**Comment by bwaidelich** [comment]
> It seems like the Neos 2.3 setup is broken without a fix for this

Nah, that was something different (see https://github.com/neos/setup/pull/13).
One way out of this chaos might be a `ComponentChain` that is initiated very early in the boot process (it could even replace the default HTTP handling IMO) and then we could always refer to that to determine the current request/response. Not the `Bootstrap` or some `RequestHandler` or...

**Comment by albe** [comment]
But isn't the ComponentChain very specific to http? So everyone would need to check if in Web or CLI context for example?!

**Comment by bwaidelich** [comment]
> But isn't the ComponentChain very specific to http?

Yes, it is (it's also in the Http sub-namespace)

> So everyone would need to check if in Web or CLI context for example?!

Do you have an example for what you mean with "everyone"?

In Web context instead of doing this:
```php
$requestHandler = $this->bootstrap->getActiveRequestHandler();
if ($requestHandler instanceof HttpRequestHandlerInterface) {
  $request = $requestHandler->getHttpRequest();
  // ...
}
```
We could maybe have sth like
```php
$request = $this->componentChain->getHttpRequest();
```
(renaming `ComponentChain` to sth like `Http\RequestContext` might be worth considering).

In CLI it's a different game anyways and I don't think that we need the same flexibility there (a component chain, ..)

**Comment by kitsunet** [comment]
Just reading up on this again. I think we should remove any notion of Request/Response from Cookie. We can inject the necessary information from the request and use a component to set whatever we need to the response.

**Comment by albe** [comment]
Just stumbled over this again and I think this is something that should be tackled soonish. I'm all for moving the Cookie-stuff out of the Session and into an Component. That requires moving the resuming of the session from the boot scripts and into the component though, so this will be a kind-of breaking change, because the Session will then only be ready after that Component.

**Comment by albe** [comment]
I guess this can be closed now with #761

**Comment by kitsunet** [comment]
Haha, magic! I didn't remember this at all, but still did it that way.
